### PR TITLE
TYP: use typing.final in indexes.base

### PR DESCRIPTION
**Pull request by jbrockmendel**
cc @simonjayhawkins this imports final conditionally following how you did it with Literal

### CODE REVIEW

File: pandas/core/indexes/base.py [hunk 1]
```diff
@@ -100,7 +100,12 @@
 )
 
 if TYPE_CHECKING:
+    from typing import final
```
Comment (jreback) [inline pandas/core/indexes/base.py:103]: can you instead define this in _typing?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexes/base.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/indexes/base.py:103]: huh, i thought i couldnt bc of the static-checking thing, but it looks like i can.  will update

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm. @simonjayhawkins if comments.

**Comment by simonjayhawkins** [comment]
Thanks @jbrockmendel lgtm

**Pull request merged**